### PR TITLE
Add unit utilities

### PR DESCRIPTION
**Pull request by Hdt80bro**
As you might have guessed from the branch name, this is in preparation for reworking the sacrifice bonus (specifically because the engine never intended for support commanders to have presets), but these utilities are probably nice to have in general. There are three main changes:
* Abstracts the total mass cost calculation out of the veterancy code into `Unit:GetTotalResourceCosts()`. It also appears that the previous calculation did not take enhancement prerequisites into account, so that's been fixed.
* Adds the concept of a "default work order," explained in the comments as "something that an assisting unit would be able to directly do for the unit without inheriting work: constructing, enhancing, upgrading, factory building, repairing, or nothing (in that order)"--these are the actions that can happen when the sacrifice command is given. Maybe I should have included that in the comments too
  * This excludes silo missile building, mobile building, shield assisting, reclaiming, capturing, and teleporting
* Provides a method to calculate the progress bonus a sacrifice will have. Note that it doesn't cap the returned progress to the actual bonus that will happen at that given moment for the current default work order. Leaving that to the user increases robustness and is more useful (and I may have plans for the progress spilling over into the next default work order when the sacrifice bonus is too large...)

### CODE REVIEW

File: lua/sim/Unit.lua [hunk 1]
```diff
@@ -6,37 +6,41 @@
...
--- cached categories for performance
 local UpdateAssistersConsumptionCats = categories.REPAIR - categories.INSIGNIFICANTUNIT     -- anything that repairs but insignificant things, such as drones
 
--- upvalue for performance
-local rawget = rawget 
-local IsAlly = IsAlly
-local armies = ListArmies()
+
+
+local DeprecatedWarnings = { }
```
Comment (Garanas) [inline lua/sim/Unit.lua:43]: This one should be unused, and can be removed

File: lua/sim/Unit.lua [hunk 2]
```diff
@@ -2622,12 +2671,16 @@ Unit = Class(moho.unit_methods) {
         end
     end,
 
-    OnStartSacrifice = function(self, target_unit)
-        EffectUtilities.PlaySacrificingEffects(self, target_unit)
+    OnStartSacrifice = function(self, targetUnit)
+        self:SetUnitState("Sacrificing", true)
+        EffectUtilities.PlaySacrificingEffects(self, targetUnit)
+        self.SacrificeThread = self:ForkThread(self.DoSacrificeThread, targetUnit)
```
Comment (Garanas) [inline lua/sim/Unit.lua:2677]: The function `DoSacrificeThread` doesn't exist

**Review by Garanas — APPROVED** [review]
Looks good - works as intended. It does feel like the pull request is not complete. Yet, it doesn't cause an error so I'm not entirely sure what is going on 😄

File: lua/sim/Unit.lua (hunk 2 above)
Comment (Hdt80bro) [reply lua/sim/Unit.lua:2677]: Doh! I forgot to remove that

**Comment by Hdt80bro** [comment]
Yeah, it's just groundwork for my future plans to rework sacrificing, so it doesn't add any functionality (besides fixing the total unit cost not including prerequisite enhancements). I thought I'd try to keep things small, since I usually struggle with that.

File: lua/sim/Unit.lua (hunk 2 above)
Comment (Garanas) [reply lua/sim/Unit.lua:2677]: Yes, but it also appears this function didn't get called. Are you sure it gets called when a unit starts sacrificing?

Comment (Hdt80bro) [reply lua/sim/Unit.lua:2677]: Yes, I was testing a while back and `OnStartSacrifice` works fine--the effects still play and everything. I actually started to implement my rework (inspiring https://github.com/FAForever/fa/pull/4069), but was able to scale it back to something manageable. I'll continue again once these two get closed.

File: lua/sim/Unit.lua (hunk 1 above)
Comment (Hdt80bro) [reply lua/sim/Unit.lua:43]: Not quite--there's still a couple uncommented out deprecation warnings in the class.

**Comment by Garanas** [comment]
I'm fine with merging this - do as you see fit @Hdt80bro

**Pull request merged**